src/redis/fixed_cache_impl.go

```diff
@@ -51,29 +59,96 @@ func (this *fixedRateLimitCacheImpl) DoLimit(
 
 	isOverLimitWithLocalCache := make([]bool, len(request.Descriptors))
 	results := make([]uint32, len(request.Descriptors))
-	var pipeline, perSecondPipeline Pipeline
+	currentCount := make([]uint32, len(request.Descriptors))
+	var pipeline, perSecondPipeline, pipelineToGet, perSecondPipelineToGet Pipeline
 
-	hitAddendForRedis := hitsAddend
+	hitsAddendForRedis := hitsAddend
 	overlimitIndex := -1
-	// Now, actually setup the pipeline, skipping empty cache keys.
-	for i, cacheKey := range cacheKeys {
-		if cacheKey.Key == "" {
-			continue
-		}
+	nearlimitIndex := -1
+	isCacheKeyOverlimit := false
+
+	if rlSettings.StopCacheKeyIncrementWhenOverlimit {
+		// Check if any of the keys are reaching to the over limit in redis cache.
+		for i, cacheKey := range cacheKeys {
+			if cacheKey.Key == "" {
+				continue
+			}
 
-		// Check if key is over the limit in local cache.
-		if this.baseRateLimiter.IsOverLimitWithLocalCache(cacheKey.Key) {
-			if limits[i].ShadowMode {
-				logger.Debugf("Cache key %s would be rate limited but shadow mode is enabled on this rule", cacheKey.Key)
+			// Check if key is over the limit in local cache.
+			if this.baseRateLimiter.IsOverLimitWithLocalCache(cacheKey.Key) {
+				if limits[i].ShadowMode {
+					logger.Debugf("Cache key %s would be rate limited but shadow mode is enabled on this rule", cacheKey.Key)
+				} else {
+					logger.Debugf("cache key is over the limit: %s", cacheKey.Key)
+				}
+				isOverLimitWithLocalCache[i] = true
+				hitsAddendForRedis = 0
+				overlimitIndex = i
+				isCacheKeyOverlimit = true
+				continue
 			} else {
-				logger.Debugf("cache key is over the limit: %s", cacheKey.Key)
+				if this.perSecondClient != nil && cacheKey.PerSecond {
+					if perSecondPipelineToGet == nil {
+						perSecondPipelineToGet = Pipeline{}
+					}
+					pipelineAppendtoGet(this.perSecondClient, &perSecondPipelineToGet, cacheKey.Key, &currentCount[i])
+				} else {
+					if pipelineToGet == nil {
+						pipelineToGet = Pipeline{}
+					}
+					pipelineAppendtoGet(this.client, &pipelineToGet, cacheKey.Key, &currentCount[i])
+				}
+			}
+		}
+
+		// Only if none of the cache keys are over the limit, call Redis to check whether cache keys are getting overlimited.
+		if len(cacheKeys) > 1 && !isCacheKeyOverlimit {
+			if pipelineToGet != nil {
+				checkError(this.client.PipeDo(pipelineToGet))
+			}
+			if perSecondPipelineToGet != nil {
+				checkError(this.perSecondClient.PipeDo(perSecondPipelineToGet))
+			}
+
+			for i, cacheKey := range cacheKeys {
+				if cacheKey.Key == "" {
+					continue
+				}
+				// Now fetch the pipeline.
+				limitBeforeIncrease := currentCount[i]
+				limitAfterIncrease := limitBeforeIncrease + hitsAddend
+
+				limitInfo := limiter.NewRateLimitInfo(limits[i], limitBeforeIncrease, limitAfterIncrease, 0, 0)
+
+				if this.baseRateLimiter.IsOverLimitThresholdReached(limitInfo) {
+					hitsAddendForRedis = 0
+					nearlimitIndex = i
```

Comment:
Seems it only handles the case when there's only one over-the-limit descriptor. If there are multiple over-the-limit descriptors at the same time, will it continue to increment the counter in other over-the-limit descriptors?

Reply:
Here if it has multiple discriptors which has same thresholds, if first one in the list is identified as it in the nearlimit, it is stop incrementing other descriptors which including the other descriptor which has same threshold also.

That mean 1 is remaining for the other descriptor which has same threshold although one we identified as in the nearlimit is getting overlimit in next request. To keep the consistency, will add a list to store and handle that same threshold scenario for multiple descriptors.

Reply:
Handled multiple key scenario in 72b9230f584b5e1357d6cb4bea1f9aac782d8a50